capellambse/extensions/validation/_validate.py

```diff
@@ -0,0 +1,483 @@
+    object type the rule is applicable to. The type will be used to feed
+    the validator with inputs (instances of type) during validation.
+    """
+    try:
+        VALIDATION_RULES.by_id(id)
+        LOGGER.warning(
+            "Failed to register rule: %r with ID: %r. Rule ID already exists",
+            name,
+            id,
+        )
+        unique_check = False
+    except KeyError:
+        unique_check = True
+
+    if not types:
+        raise TypeError("No 'types' specified")
+    if isinstance(types, type):
+        type_names = [types.__name__]
+    elif isinstance(types, (RealType, VirtualType)):
+        type_names = [types.name]
+    elif isinstance(types, str):
+        type_names = [types]
+    else:
+        type_names = []
+        for i in types:
+            if isinstance(i, str):
+                type_names.append(i)
+            elif isinstance(i, (RealType, VirtualType)):
+                type_names.append(i.name)
+            else:
+                type_names.append(i.__name__)
+
+    def rule_decorator(
+        rule: cabc.Callable[[_T], bool]
+    ) -> cabc.Callable[[_T], bool]:
+        rule = Rule(
+            id,
+            name,
+            type_names,
+            rationale,
+            category,
+            action,
+            rule,
+            hyperlink_further_reading,
+        )
+        if unique_check:
+            VALIDATION_RULES[category].append(rule)
+        return rule
+
+    return rule_decorator
+
+
+def store_result(rule: Rule, obj: c.GenericElement, result: Result) -> None:
+    """Save a ``result`` for the given ``rule`` and ModelObject."""
+    results = obj._model.validation.results
+    results.setdefault(rule, {}).setdefault(obj.uuid, result)
+
+
+class Validation:
+    """Basic class for access to validation rules and results."""
+
+    _model: capellambse.MelodyModel
+    _element: c.GenericElement
+
+    def __init__(self, **kw: t.Any) -> None:
+        raise TypeError("Cannot create Validation object this way")
+
+    @classmethod
+    def from_model(
+        cls, model_: capellambse.MelodyModel, element: etree._Element
+    ) -> Validation:
+        """Create a Validation object for a MelodyModel or an element."""
+        self = cls.__new__(cls)
+        self._model = model_
+        self._element = c.GenericElement.from_model(model_, element)
+        return self
+
+
+class ModelValidation(Validation):
+    """Provides access to the model's validation rules and results."""
+
+    @property
+    def rules(self) -> Rules:
+        """Return all registered validation rules."""
+        return VALIDATION_RULES
+
+    @property
+    def results(self) -> Results:
+        """Return all stored validation rule results."""
+        key = f"{__name__}.RESULTS"
+        if key not in self._model.__dict__:
+            self._model.__dict__[key] = Results()
+            self.validate()
+        return self._model.__dict__[key]
```

Comment:
I just realized that my first point (uninitialized Results object) wouldn't actually happen, because on first access validation does happen; however, this also means that an explicit call to `validate()` performs the full model validation twice. This sort of logic error obviously wouldn't have been possible either.